File: 0098/main.py
from typing import Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        nodes = deque()
        
        nodes.append(root)
        
        tmp = root.left
        while tmp != None:
            nodes.append(tmp)
            tmp = tmp.left  
        
        previous_val = None
        while len(nodes) > 0:
            curr = nodes.pop()
            if previous_val is not None and previous_val >= curr.val:
                return False
            
            previous_val = curr.val
            
            # add right + all left children
            if curr.right is not None:                
                tmp = curr.right
                
                while tmp is not None:
                    nodes.append(tmp)
                    tmp = tmp.left
        
        return True

if __name__ == "__main__":
    pass

File: 0098/main.cpp
int main() {

}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include <stack>

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        std::stack<TreeNode*> s;
        
        TreeNode* curr = root;
        while (curr != NULL) {
            s.push(curr);
            curr = curr->left;
        }
        
        int* prev = NULL;
        while (!s.empty()) {
            curr = s.top();
            s.pop();
            
            if (prev && *prev >= curr->val) {
                return false;
            }
            
            prev = &(curr->val);            
            curr = curr->right;
            while (curr) {
                s.push(curr);
                curr = curr->left;
            }
        }
        
        return true;
    }
};